File: src/types/review.ts
import type { Course, Place, ReviewMood } from './core';

export interface Comment {
  id: string;
  userId: string;
  author: string;
  body: string;
  parentId: string | null;
  isDeleted: boolean;
  createdAt: string;
  replies: Comment[];
}

export interface Review {
  id: string;
  userId: string;
  placeId: string;
  placeName: string;
  author: string;
  body: string;
  mood: ReviewMood;
  badge: string;
  visitedAt: string;
  imageUrl: string | null;
  thumbnailUrl?: string | null;
  commentCount: number;
  likeCount: number;
  likedByMe: boolean;
  stampId: string | null;
  visitNumber: number;
  visitLabel: string;
  travelSessionId: string | null;
  hasPublishedRoute: boolean;
  comments: Comment[];
}

export interface ReviewFeedPageResponse {
  items: Review[];
  nextCursor: string | null;
}

export interface StampLog {
  id: string;
  placeId: string;
  placeName: string;
  stampedAt: string;
  stampedDate: string;
  visitNumber: number;
  visitLabel: string;
  travelSessionId: string | null;
  travelSessionStampCount: number;
  isToday: boolean;
}

export interface TravelSession {
  id: string;
  startedAt: string;
  endedAt: string;
  durationLabel: string;
  stampCount: number;
  placeIds: string[];
  placeNames: string[];
  canPublish: boolean;
  publishedRouteId: string | null;
  coverPlaceId: string | null;
}

export interface ReviewLikeResponse {
  reviewId: string;
  likeCount: number;
  likedByMe: boolean;
}

export interface UserRoute {
  id: string;
  authorId: string;
  author: string;
  title: string;
  description: string;
  mood: string;
  likeCount: number;
  likedByMe: boolean;
  createdAt: string;
  placeIds: string[];
  placeNames: string[];
  isUserGenerated: boolean;
  travelSessionId: string | null;
}

export interface UserRouteLikeResponse {
  routeId: string;
  likeCount: number;
  likedByMe: boolean;
}

export interface StampState {
  collectedPlaceIds: string[];
  logs: StampLog[];
  travelSessions: TravelSession[];
}

export interface BootstrapResponse {
  auth: import('./auth').AuthSessionResponse;
  places: Place[];
  reviews: Review[];
  courses: Course[];
  stamps: StampState;
  hasRealData: boolean;
}

export interface MapBootstrapResponse {
  auth: import('./auth').AuthSessionResponse;
  places: Place[];
  stamps: StampState;
  hasRealData: boolean;
}

export interface CourseBootstrapResponse {
  courses: Course[];
}

export interface ReviewCreateRequest {
  placeId: string;
  stampId: string;
  body: string;
  mood: ReviewMood;
  imageUrl?: string | null;
}

export interface ReviewUpdateRequest {
  body: string;
  mood: ReviewMood;
  imageUrl?: string | null;
}

export interface CommentCreateRequest {
  body: string;
  parentId?: string | null;
}

export interface UserRouteCreateRequest {
  title: string;
  description: string;
  mood: string;
  travelSessionId: string;
  isPublic?: boolean;
}

export interface StampClaimRequest {
  placeId: string;
  latitude: number;
  longitude: number;
}

File: backend/app/model_contracts/review.py
"""Review and comment API models."""

from __future__ import annotations

from pydantic import Field

from .core import ApiModel, ReviewMood


class CommentOut(ApiModel):
    id: str
    user_id: str = Field(alias='userId')
    author: str
    body: str
    parent_id: str | None = Field(default=None, alias='parentId')
    is_deleted: bool = Field(alias='isDeleted')
    created_at: str = Field(alias='createdAt')
    replies: list['CommentOut'] = Field(default_factory=list)


class ReviewOut(ApiModel):
    id: str
    user_id: str = Field(alias='userId')
    place_id: str = Field(alias='placeId')
    place_name: str = Field(alias='placeName')
    author: str
    body: str
    mood: ReviewMood | str
    badge: str
    visited_at: str = Field(alias='visitedAt')
    image_url: str | None = Field(default=None, alias='imageUrl')
    thumbnail_url: str | None = Field(default=None, alias='thumbnailUrl')
    comment_count: int = Field(alias='commentCount')
    like_count: int = Field(default=0, alias='likeCount')
    liked_by_me: bool = Field(default=False, alias='likedByMe')
    stamp_id: str | None = Field(default=None, alias='stampId')
    visit_number: int = Field(default=1, alias='visitNumber')
    visit_label: str = Field(alias='visitLabel')
    travel_session_id: str | None = Field(default=None, alias='travelSessionId')
    has_published_route: bool = Field(default=False, alias='hasPublishedRoute')
    comments: list[CommentOut] = Field(default_factory=list)


class ReviewLikeResponse(ApiModel):
    review_id: str = Field(alias='reviewId')
    like_count: int = Field(alias='likeCount')
    liked_by_me: bool = Field(alias='likedByMe')


class ReviewCreate(ApiModel):
    place_id: str = Field(alias='placeId')
    stamp_id: str = Field(alias='stampId')
    body: str
    mood: ReviewMood | str
    image_url: str | None = Field(default=None, alias='imageUrl')


class CommentCreate(ApiModel):
    body: str
    parent_id: str | None = Field(default=None, alias='parentId')
